crates/guest-rust/rt/src/async_support/stream_support.rs

```diff
@@ -1,133 +1,187 @@
+    /// The raw `stream.close-readable` intrinsic.
+    pub close_readable: unsafe extern "C" fn(stream: u32),
+    /// The raw `stream.new` intrinsic.
+    pub new: unsafe extern "C" fn() -> u64,
 }
 
 /// Helper function to create a new read/write pair for a component model
 /// stream.
 pub unsafe fn stream_new<T>(
     vtable: &'static StreamVtable<T>,
 ) -> (StreamWriter<T>, StreamReader<T>) {
-    let handle = unsafe { (vtable.new)() };
-    super::with_entry(handle, |entry| match entry {
-        Entry::Vacant(entry) => {
-            entry.insert(Handle::LocalOpen);
-        }
-        Entry::Occupied(_) => unreachable!(),
-    });
-    (
-        StreamWriter::new(handle, vtable),
-        StreamReader::new(handle, vtable),
-    )
-}
-struct CancelWriteOnDrop<T: 'static> {
-    handle: Option<u32>,
-    vtable: &'static StreamVtable<T>,
-}
-
-impl<T> Drop for CancelWriteOnDrop<T> {
-    fn drop(&mut self) {
-        if let Some(handle) = self.handle.take() {
-            super::with_entry(handle, |entry| match entry {
-                Entry::Vacant(_) => unreachable!(),
-                Entry::Occupied(mut entry) => match entry.get() {
-                    Handle::LocalOpen
-                    | Handle::LocalWaiting(_)
-                    | Handle::Read
-                    | Handle::LocalClosed
-                    | Handle::WriteClosedErr(_) => unreachable!(),
-                    Handle::LocalReady(..) => {
-                        entry.insert(Handle::LocalOpen);
-                    }
-                    Handle::Write => unsafe {
-                        // TODO: spec-wise this can return `BLOCKED` which seems
-                        // bad?
-                        (self.vtable.cancel_write)(handle);
-                    },
-                },
-            });
-        }
+    unsafe {
+        let handles = (vtable.new)();
+        let writer = handles as u32;
+        let reader = (handles >> 32) as u32;
+        rtdebug!("stream.new() = [{writer}, {reader}]");
+        (
+            StreamWriter::new(writer, vtable),
+            StreamReader::new(reader, vtable),
+        )
     }
 }
 
 /// Represents the writable end of a Component Model `stream`.
 pub struct StreamWriter<T: 'static> {
     handle: u32,
-    future: Option<Pin<Box<dyn Future<Output = ()> + 'static>>>,
     vtable: &'static StreamVtable<T>,
 }
 
 impl<T> StreamWriter<T> {
     #[doc(hidden)]
-    pub fn new(handle: u32, vtable: &'static StreamVtable<T>) -> Self {
-        Self {
-            handle,
-            future: None,
-            vtable,
+    pub unsafe fn new(handle: u32, vtable: &'static StreamVtable<T>) -> Self {
+        Self { handle, vtable }
+    }
+
+    /// Initiate a write of the `values` provided into this stream.
+    ///
+    /// This method is akin to an `async fn` except that the returned
+    /// [`StreamWrite`] future can also be cancelled via [`StreamWrite::cancel`]
+    /// to re-acquire intermediate values.
```

Comment:
```suggestion
    /// to re-acquire undelivered values.
```
This is just a matter of opinion, but "undelivered" is more meaningful to me in this context than "intermediate".